handler/rfc8693/handler.go

```diff
@@ -0,0 +1,269 @@
+	//		in the actor_token parameter. This is REQUIRED when the actor_token parameter is present
+	//		in the request but MUST NOT be included otherwise.
+	actorTokenType := form.Get("actor_token_type")
+	if actorTokenType != "" {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("'actor_token_type' was provided but delegation is currently not supported."))
+	}
+
+	return &tokenExchangeParams{
+		subjectToken:     subjectToken,
+		subjectTokenType: subjectTokenType,
+	}, nil
+}
+
+func (c *Handler) HandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(requester) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	client := requester.GetClient()
+	if client.GetID() == "" {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("unauthenticated client"))
+	}
+
+	// Check whether client is allowed to use token exchange.
+	if !client.GetGrantTypes().Has(string(fosite.GrantTypeTokenExchange)) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("the client is not allowed to use token-exchange"))
+	}
+
+	// Get request parameter related token exchange.
+	params, err := parseRequestParameter(requester)
+	if err != nil {
+		return err
+	}
+
+	// Check and grant scope.
+	for _, scope := range requester.GetRequestedScopes() {
+		if !c.Config.GetScopeStrategy(ctx)(client.GetScopes(), scope) {
+			return errorsx.WithStack(fosite.ErrInvalidScope.WithHintf("The OAuth 2.0 Client is not allowed to request scope '%s'.", scope))
+		}
+		requester.GrantScope(scope)
+	}
+
+	// Check and grant audience.
+	if err := c.Config.GetAudienceStrategy(ctx)(client.GetAudience(), requester.GetRequestedAudience()); err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("audience not match: %v", err))
+	}
+	for _, audience := range requester.GetRequestedAudience() {
+		requester.GrantAudience(audience)
+	}
+
+	// Verify subject token.
+	switch params.subjectTokenType {
+	case tokenTypeIDToken:
+		claims := jwt.MapClaims{}
+		if _, err := jwt.ParseWithClaims(params.subjectToken, claims, c.keyFunc(ctx)); err != nil {
+			return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("failed to verify JWT: %v", err))
+		}
+		subject, err := c.Storage.GetImpersonateSubject(ctx, claims, requester)
+		if err != nil {
+			return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("not allowed to token exchange by jwt: %v", err))
+		}
+		requester.SetSession(&fosite.DefaultSession{
+			Subject: subject,
+		})
+		requester.GetSession().SetExpiresAt(fosite.AccessToken, time.Now().UTC().Add(c.HandleHelper.Config.GetAccessTokenLifespan(ctx)))
+		return nil
+	case tokenTypeAT:
+		or, err := c.verifyAccessTokenAsSubjectToken(ctx, client.GetID(), params)
+		if err != nil {
+			return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("not allowed to token exchange by at: %v", err))
+		}
+		requester.SetSession(or.GetSession().Clone())
```

Comment:
I am not sure we should clone the subject token's session here because it might contain extra claims that have nothing to do with the token requested. Also, wouldn't this modify the client_id, aud and other claims in the introspection response (maybe it doesn't)?

A more controlled change might be to introduce a Session interface for this package and add a `SetSubject(string)` to it.